src/MetaMorpho.sol

```diff
@@ -326,20 +336,23 @@ contract MetaMorpho is ERC4626, ERC20Permit, Ownable2Step, Multicall, IMetaMorph
             seen[prevIndex] = true;
 
             newWithdrawQueue[i] = id;
-
-            // Safe "unchecked" cast because i < currLength.
-            config[id].withdrawRank = uint64(i + 1);
         }
 
         for (uint256 i; i < currLength; ++i) {
             if (!seen[i]) {
                 Id id = withdrawQueue[i];
 
-                if (MORPHO.supplyShares(id, address(this)) != 0 || config[id].cap != 0) {
-                    revert ErrorsLib.InvalidMarketRemoval(id);
+                if (config[id].cap != 0) revert ErrorsLib.InvalidMarketRemovalNonZeroCap(id);
+
+                if (config[id].removableAt != 0) {
+                    if (block.timestamp < config[id].removableAt) {
+                        revert ErrorsLib.InvalidMarketRemovalTimelockNotElapsed(id);
+                    }
+                } else if (MORPHO.supplyShares(id, address(this)) != 0) {
+                    revert ErrorsLib.InvalidMarketRemovalNonZeroSupply(id);
```

Comment:
```suggestion
                if (config[id].removableAt != 0 && block.timestamp < config[id].removableAt) {
                        revert ErrorsLib.InvalidMarketRemovalTimelockNotElapsed(id);
                    }
                } else if (MORPHO.supplyShares(id, address(this)) != 0) {
                    revert ErrorsLib.InvalidMarketRemovalNonZeroSupply(id);
```

Reply:
The suggested behavior is not the same and is faulty: if some market is removable (the timelock timestamp has passed), then the suggested code requires the removable market to have 0 supply shares

Reply:
You're right